### PR TITLE
feat: resolve artifacts by their content hash

### PR DESCRIPTION
**Pull request by gr8routdoors**
This addresses Issue #3237, adding the ability to resolve artifacts by their content hash in the Kafka serde. Note that resolving a schema by content hash directly from their data in `DefaultSchemaResolver` is not supported, as the schema data is generated in these cases and has a very good chance of not being consistent with the schema content that was actually registered. Not only would this cause poor results, but also cost a server call and worsen performance.

### CODE REVIEW

**Comment by gr8routdoors** [comment]
Note that I tried looking to see if there was an integration test that covered this type of behavior for other types of IDs, but was unable to find one.  Please point me in the wrong direction if one exists and I missed.

**Comment by EricWittmann** [comment]
The build is reporting a compile issue - does the maven build work locally for you?

**Comment by gr8routdoors** [comment]
Updated to current main, all tests passing.  Should be good for review @EricWittmann and @carlesarnal

**Review by carlesarnal — COMMENTED** [review]
Looks good to me. The only "concern" I have is that this will be hard to use for artifacts with references since those are used now for the contenthash calculations. That said, I think it's probably fine to have the option. Thanks for the contribution!

**Comment by carlesarnal** [comment]
@smccarthy-ie we need to add some documentation here for this new option.

File: schema-resolver/src/main/java/io/apicurio/registry/resolver/SchemaLookupResult.java [hunk 1]
```diff
@@ -82,6 +90,7 @@ public ArtifactReference toArtifactReference() {
         return ArtifactReference.builder()
                 .globalId(this.getGlobalId())
                 .contentId(this.getContentId())
+                .contentHash(this.contentHash)
```
[the remark this answers is not in the document]
Comment (gr8routdoors) [reply schema-resolver/src/main/java/io/apicurio/registry/resolver/SchemaLookupResult.java]: Yep fixed that in the latest commit

File: schema-resolver/src/test/java/io/apicurio/registry/resolver/AbstractSchemaResolverTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 Red Hat
```
[the remark this answers is not in the document]
Comment (gr8routdoors) [reply schema-resolver/src/test/java/io/apicurio/registry/resolver/AbstractSchemaResolverTest.java]: Fixed the copyright dates in all the new files.

**Comment by gr8routdoors** [comment]
> Looks good to me. The only "concern" I have is that this will be hard to use for artifacts with references since those are used now for the contenthash calculations. That said, I think it's probably fine to have the option. Thanks for the contribution!

Sure thing!  I was worried about that too when I saw [your PR](https://github.com/Apicurio/apicurio-registry/pull/3227).  However, we're using protobuf, which it looks like [doesn't consider references](https://github.com/Apicurio/apicurio-registry/blob/main/schema-util/protobuf/src/main/java/io/apicurio/registry/content/canon/ProtobufContentCanonicalizer.java#L38-L47) when it [canonicalizes content to calculate the canonical content hash](https://github.com/Apicurio/apicurio-registry/blob/main/app/src/main/java/io/apicurio/registry/storage/impl/sql/upgrader/ProtobufCanonicalHashUpgrader.java#L69-L96), so I'm hoping it will still work for us. 🤞

**Comment by EricWittmann** [comment]
Thanks @gr8routdoors !

**Pull request merged**